### PR TITLE
XPlat fixes for WebAssembly.

### PR DESCRIPTION
**Pull request by Penguinwizzard**
A few changes to help with WebAssembly on xplat. Note that while most tests pass, there's still a couple of issues: firstly, liveness of entrypoints in the debugger.js test is incorrectly handled, and secondly, we uncovered a bug that also happens in asm.js related to floating-point arguments. Due to this, ENABLE_WASM isn't set on xplat yet, but chakracore does now build and run with that flag set.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/2381)

### CODE REVIEW

File: lib/WasmReader/WasmSignature.cpp [hunk 1]
```diff
@@ -181,6 +181,54 @@ WasmSignature::FromIDL(WasmSignatureIDL* sig)
     return reinterpret_cast<WasmSignature*>(sig);
 }
 
+void
+WasmSignature::Dump()
+{
+    Output::Print(_u("("));
+    for(uint32 i=0; i<this->GetParamCount(); i++)
```
Comment (obastemur) [inline lib/WasmReader/WasmSignature.cpp]: nit: spaces between operators. (i.e. i = 0 .. i < .. )

File: lib/Runtime/Library/CMakeLists.txt [hunk 2]
```diff
@@ -114,6 +114,13 @@ set(CRLIB_SOURCE_CODES
     TypedArray.cpp
     TypedArrayIndexEnumerator.cpp
     UriHelper.cpp
+    WasmLibrary.cpp
```
Comment (obastemur) [inline lib/Runtime/Library/CMakeLists.txt]: I would really appreciate if you can add these sources conditionally. 

```
if (CC_TARGETS_AMD64)
    set(CRLIB_SOURCE_CODES ${CRLIB_SOURCE_CODES}
        WasmLibrary.cpp
        WebAssembly.cpp
        .... and others
       )
endif()
```

..and an `xplat-todo` note `enable this for other arch(s)` same applies to other CMake etc. changes.

**Comment by Cellule** [comment]
Reviewed 1 of 4 files at r1, 12 of 13 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[lib/WasmReader/WasmSignature.cpp, line 187 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/2381#-KaoGzvhDwa_Ar_8mufW:-KaoGzvhDwa_Ar_8mufX:b-jg5biz) ([raw file](https://github.com/microsoft/chakracore/blob/17416390793db6d8bf9d913c515b755bffca7fc8/lib/WasmReader/WasmSignature.cpp#L187)):*
> ```cpp
> WasmSignature::Dump()
> {
>     Output::Print(_u("("));
> ```

Could you wrap the content of this function with `#if ENABLE_DEBUG_CONFIG_OPTIONS`

---

*[test/WasmSpec/rlexe.xml, line 2 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/2381#-KaoHKNlE5VHv0Gw15Z4:-KaoHKNlE5VHv0Gw15Z5:b7j5ox3) ([raw file](https://github.com/microsoft/chakracore/blob/17416390793db6d8bf9d913c515b755bffca7fc8/test/WasmSpec/rlexe.xml#L2)):*
> ```XML
> <?xml version="1.0" encoding="utf-8"?>
> 
> ```

Like the comment mentions, this file is automatically generated, thus you should also edit the source that creates this file.
Change `<baseline>${getBaselinePath(specFile)}</baseline>` in  `test\WasmSpec\convert-test-suite\index.js` to replace backward slash to forward slash

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/2381)*

**Comment by Cellule** [comment]
@dotnet test OSX osx_osx_debug_static please

**Comment by Cellule** [comment]
Reviewed 2 of 3 files at r3, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[lib/Runtime/Library/CMakeLists.txt, line 117 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/2381#-KaoGMf7BvAKxPLBM-XX-r2-117:-KbRO9WFDjJlGlehxFol:b6jc9kp) ([raw file](https://github.com/microsoft/chakracore/blob/17416390793db6d8bf9d913c515b755bffca7fc8/lib/Runtime/Library/CMakeLists.txt#L117)):*
<details><summary><i>Previously, obastemur (Oguz Bastemur) wrote…</i></summary><blockquote>

I would really appreciate if you can add these sources conditionally. 

```
if (CC_TARGETS_AMD64)
    set(CRLIB_SOURCE_CODES ${CRLIB_SOURCE_CODES}
        WasmLibrary.cpp
        WebAssembly.cpp
        .... and others
       )
endif()
```

..and an `xplat-todo` note `enable this for other arch(s)` same applies to other CMake etc. changes.
</blockquote></details>

@penguinwizzard Can you look at this request ?

---

*[test/WasmSpec/convert-test-suite/index.js, line 95 at r4](https://reviewable.io:443/reviews/microsoft/chakracore/2381#-KbRNnRP9OYoJEWGdQ8b:-KbRNnRP9OYoJEWGdQ8c:b-ltuqjk) ([raw file](https://github.com/microsoft/chakracore/blob/29c5752e6ceb9eddb196524ac5d0d8fd2ccfde18/test/WasmSpec/convert-test-suite/index.js#L95)):*
> ```JavaScript
> function hostFlags(specFile, {useFullpath} = {}) {
>   return `-wasm -args ${
>     useFullpath ? specFile : path.relative(rlRoot, specFile)
> ```

I think you should call `slash` here as well. our host is probably able to support backward slash in arguments, but might as well keep them conform across the file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/2381)*

**Comment by obastemur** [comment]
@dotnet test OSX osx_osx_debug_static please

**Comment by obastemur** [comment]
in case the CI is just fine #2405 should take care of the random initialization failures.

File: lib/WasmReader/WasmSignature.cpp (hunk 1 above)
Comment (Penguinwizzard) [inline lib/WasmReader/WasmSignature.cpp]: Fixed.

File: lib/Runtime/Library/CMakeLists.txt (hunk 2 above)
Comment (Penguinwizzard) [inline lib/Runtime/Library/CMakeLists.txt]: I've added this and Jenkins is green, sending it in in this state as per out-of-band discussion.

**Comment by Penguinwizzard** [comment]
(Since reviewable is crashing when I try to submit these responses there:)
I've added the slash command to the second location in the conversion script, and re-generated the rlexe file from it as appropriate.

The cmake files have additions to only include the wasm-related files if building on amd64 for xplat builds.

**Comment by Cellule** [comment]
:lgtm: on my side
@obastemur  Could you review the changes in cmakelist files.

---

Reviewed 6 of 6 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/2381#-:-KbSThemJCxXu6Lc5m4D:b-4j3ax8)*

**Comment by obastemur** [comment]
:shipit:

**Pull request merged**